### ISSUE TITLE
Multiple GPUs

### ISSUE THREAD

**Issue by gitkol** [choderalab/openmmtools#726]
Hi,

Can rest.py run on multiple GPUs?

Thanks,

Istvan

**Comment by ijpulidos** [comment]
Hello! What do you mean by `rest.py`? Can you be more specific as to what your issue is?

**Comment by gitkol** [comment]
I am sorry, I mean the openmmtools script for replica exchange solute
tempering. It works very well, but since every replica is running on the
same GPU, it is slow. I was wondering if workstations with more than one
GPU could be used to distribute the replicas. It is easy to run OpenMM jobs
in such manner by specifying multiple device indices for the simulation
platform, but I couldn’t find a similar option within openmmtools. Hope
this makes my question more clear. Thank you very much, Istvan

**Comment by gitkol** [comment]
I posted this question first on the OpenMM cookbook GitHub but Peter
Eastman said that it belonged here.

**Comment by xiaowei-xie2** [comment]
I have the same question. I am able to run multiple solute tempering REMD simulations in parallel with mpirun according to this issue(https://github.com/choderalab/openmmtools/issues/648), but I don't know how to distribute replicas among multiple GPUs so that they contribute to the same REMD simulation.

**Comment by gitkol** [comment]
Yes, that was exactly my point. When I run such a simulation using
openmmtools, the only option seems to be that all, say, 8 jobs run on the
same GPU and that makes it prohibitively slow. I have not received any
reply from the developers. (Plain OpenMM jobs can readily run on multiple
GPUs.)

**Comment by xiaowei-xie2** [comment]
Hi @gitkol, I think I figured it out. Do you have mpi4py installed correctly? For me, what I found was that, if I don't have mpi4py installed, mpirun would run multiple copies of the same REMD (each GPU would run a whole REMD but multiple GPUs running at the same time), but when I have mpi4py installed multiple GPUs will contribute to a single REMD. Here is an example of job files that worked for me (if it's helpful). On my system, using 4GPUs resulted in 2x speed up compared to 1 GPU (not 4x). 
[test_rest_14.tar.gz](https://github.com/choderalab/openmmtools/files/15382434/test_rest_14.tar.gz)

**Comment by ijpulidos** [comment]
@xiaowei-xie2 is correct, having `mpi4py` is important in this case. Thank you for providing a test script that we can use to reproduce your results.

There's always some part of the code that cannot be fully parallelized, for example when communicating between the different GPUs. It would be interesting to see if we can accomplish some profiling to check where the overhead is. Thanks!

**Comment by xiaowei-xie2** [comment]
Hi @ijpulidos, thank you for the insight. Yes I totally understand using n GPUs won't necessarily result in n times speed up (sometimes not even any speed up), so I am actually satisfied with the current performance. But yes it would be nice to see where the overhead is!

I am also curious does the current repo support parallelizing across multiple GPUs across multiple nodes?

**Comment by ijpulidos** [comment]
@xiaowei-xie2 It does support that, since everything is handled by the MPI environment. That also means it's highly dependent on the MPI setup of the system. Depending on the connectivity of your HPC system and the system being simulated it might make sense, or not, to do this. 

We should try to come up with an example on how to accomplish this that people can use and add it to the docs.

**Comment by jakublala** [comment]
Thanks for the comments here, @ijpulidos  - following up on the above, is it possible to specify each replica to live on its own GPU somehow?

It seems to me that that setup show provide the best performance, or is that what happens by default under the hood?

**Comment by teshort** [comment]
Had this problem recently and found the following solution:

**Context**
 
- Running on HPC cluster using PBS job scheduler.
- Each node has 4 GPUs and I want to run HREX with 8 replicas in total, assigning each to its own GPU, requiring 2 nodes in total.

In the python script:

**hrex.py**

```
# required imports

import mpiplus      # (maybe not needed)
import mpi4py
from mpi4py import MPI

# rank = process ID
# e.g. if running 8 processes (1 for each replica)
# each process will have a rank from 0 to 7.

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

# each node has devices 0 to 3 so we % 4 the rank
# and assign it to a device.
# (this has to come AFTER the property object has been made).

devid = rank % 4
platform.setPropertyDefaultValue('DeviceIndex', f'{devid}')

# everything else is handled by the openmm/mpi code
# using the **distribute** function to manage the replicas.
```

In the job submission script:

**run_hrex.qsub**

```
# required modules

module load intel-mpi
module load cuda

# creates a file containing a list of the available nodes

cat ${PBS_NODEFILE} | uniq > ${hostfile}

# -np (num. of processes) should = the number of GPUS
# (each replica gets its own process and each process gets
# assigned to its own GPU).
# -ppn (processes per node) in this case 4.

mpi_args="-f ${hostfile} -np ${PBS_NGPUS} -ppn ${ppn}"

# submit hrex.py script via mpiexec loaded  from
# module above

mpiexec ${mpi_args} python hrex.py

```

Essentially all that needs to be done is to tell the platform in the hrex.py script how to distribute the work for each copy of the script in each process.

Hope this helps!
